onnxruntime/core/graph/graph_utils.cc

```diff
@@ -662,5 +661,51 @@ void FinalizeNodeFusion(Graph& graph, const std::vector<std::reference_wrapper<N
   }
 }
 
+const Node::EdgeEnd*
+GetInputEdge(const Node& node, int arg_index) {
+  for (auto it = node.InputEdgesBegin(), end = node.InputEdgesEnd(); it != end; ++it) {
+    if (arg_index == it->GetDstArgIndex()) {
+      return &(*it);
+    }
+  }
+  return nullptr;
+}
+
+const Node* GetInputNode(const Node& node, int arg_index) {
+  const Node::EdgeEnd* edge = GetInputEdge(node, arg_index);
+  if (nullptr == edge) {
+    return nullptr;
+  }
+  return &(edge->GetNode());
+}
+
+bool FindPath(const Node& node, const std::vector<MatchEdgeEnd>& edges_to_match, std::vector<const Node::EdgeEnd*>& result) {
+  result.clear();
+  result.reserve(edges_to_match.size());
+
+  const Node* current_node = &node;
+  for (const MatchEdgeEnd& edge : edges_to_match) {
+    const Node::EdgeEnd* edge_found = nullptr;
+
+    auto edges_beign = edge.is_input_edge ? node.InputEdgesBegin() : node.OutputEdgesBegin();
+    auto edges_end = edge.is_input_edge ? node.InputEdgesEnd() : node.OutputEdgesEnd();
```

Comment:
It is fine to  have input and output edge mixed. This function starts from one node, and goes to an adjacent node by matching an edge (either input or output edge). Then continue this step to match the next edge from the adjacent node.

Reply:
I believe we should not allow this.
It will not be a path in a DAG (it could be a path in a non-directed graph, but our graph is directed).
Also, you cannot guarantee that you will not return an undirected cycle this way (what if I ask for a mat->mul->mat and mat matches the same node in both cases).
So, let's specify the direction (upwards or downwards) as a parameter in the method.